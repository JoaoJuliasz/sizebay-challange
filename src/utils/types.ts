export type FilterType = 'done' | 'pending' | ''

export type Items = {
    value: string,
    status: 'done' | 'pending'
    itemIndex: number
}