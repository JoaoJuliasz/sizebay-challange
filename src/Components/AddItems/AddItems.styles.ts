import styled from 'styled-components'
import { PlusOutlined } from '@ant-design/icons'

type AddItemsButtonProps = {
    addItemsInput: string
}

export const AddItemsContainer = styled.div`
    width: 100%;
    height: calc(100% - 145px);
`

export const AddItemsInputContainer = styled.div`
    width: 80%;
    border: 1px solid #dbdbdb;
    border-radius: 5px;
    margin: auto;

    @media only screen and (max-width: 768px) {
        width: 86%;
    }
`

export const AddItemsInput = styled.input`
    min-width: calc(100% - 50px);
    padding: 12px 10px;
    outline: none;
    border-right: none;
    border-radius: 5px 0px 0px 5px;
    border: none;
`

export const AddItemsButton = styled.button<AddItemsButtonProps>`
    width: 50px;
    border: 1px solid #98c2c9;
    border-left: none;
    padding: 12px;
    border-radius: 0 5px 5px 0;
    background: #4da6b3;
    cursor: ${props => props.addItemsInput ? 'pointer' : 'not-allowed'};
    opacity: ${props => !props.addItemsInput ? 0.7 : 1}
`

export const AddItemsItemContainer = styled.div`
    width: 100%;
    margin: 10px 0;
    height: calc(100% - 75px);
    overflow: auto;
`

export const PlusIcon = styled(PlusOutlined) `
    background: #fff;
    padding: 4px;
    border-radius: 50%;
    color: #4da6b3;
`