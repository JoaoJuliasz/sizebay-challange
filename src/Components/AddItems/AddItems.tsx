import { useState } from 'react';
import { FilterType, Items } from '../../utils/types';

//Importing libs
import { toast } from 'react-toastify';

//Importing styled-components
import { AddItemsContainer, AddItemsInputContainer, AddItemsInput, AddItemsButton, AddItemsItemContainer, PlusIcon } from './AddItems.styles'

import Item from '../Item/Item';

type AddItemsProsp = {
    items: Items[]
    oldItems: Items[]
    setItems: (value: Items[]) => void
    filterType: FilterType
    setFilterType: (value: FilterType) => void
    searchField: string
    setSearchField: (value: string) => void
}

const AddItems = ({ items, oldItems, setItems, filterType, setFilterType, searchField, setSearchField }: AddItemsProsp) => {
    const [addItemsInput, setAddItemsInput] = useState<string>('')

    const handleClick = () => {
        const auxArr = [...oldItems]
        auxArr.push({ value: addItemsInput, status: 'pending', itemIndex: auxArr.length > 0 ? auxArr.length : 0 })
        setTimeout(() => {
            setItems(auxArr)
        }, 0)
        setAddItemsInput(prev => '')
    }

    const handleChange = ({target: {value}}: React.ChangeEvent<HTMLInputElement>) => {
        setAddItemsInput(value)
    }

    const changeItemStatus = (itemIndex: number, action: string) => {
        const newItems = [...oldItems]
        const index = newItems.findIndex(item => item.itemIndex === itemIndex)
        if (action === 'remove') {
            newItems.splice(index, 1)
            toast.success('Item removed')
        } else {
            newItems[index].status = 'done'
            toast.success('Item completed')
        }
        setItems(newItems)
    }

    const changeItemValue = (itemIndex: number, value: string) => {
        const newItems = items
        const index = newItems.findIndex(item => item.itemIndex === itemIndex)
        newItems[index].value = value
        setItems(newItems)
    }

    const clearFilters = () => {
        setFilterType('')
        setSearchField('')
    }

    return (
        <AddItemsContainer>
            {!(filterType || searchField) &&
                <AddItemsInputContainer className="add-item-input">
                    <AddItemsInput id="add-item-field" placeholder="Add new item" onChange={e => handleChange(e)} value={addItemsInput} type="text" />
                    <AddItemsButton id="add-item-btn" disabled={!addItemsInput} addItemsInput={addItemsInput} onClick={handleClick}>
                        <PlusIcon />
                    </AddItemsButton>
                </AddItemsInputContainer>}
            <AddItemsItemContainer>
                {items?.map((item, idx) =>
                    <Item key={idx} idx={idx} item={item} changeItemValue={changeItemValue} changeItemStatus={changeItemStatus} />
                )}
                {((filterType || searchField) && items.length === 0) &&
                    <div>
                        There are no items marked as done. <span id="clear-filter" style={{ textDecoration: 'underline', cursor: 'pointer' }} onClick={() => clearFilters()}>Clear the filter here to see all items.</span>
                    </div>
                }
            </AddItemsItemContainer>
        </AddItemsContainer>
    );
};

export default AddItems;