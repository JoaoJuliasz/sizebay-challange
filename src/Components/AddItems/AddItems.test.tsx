import React from 'react';
import { shallow } from 'enzyme'
import AddItems from './AddItems'
import { FilterType, Items } from '../../utils/types';

describe('AddItems', () => {

    const mockItems = (items: Items[], filterType: FilterType, searchField: string) => {
        return items.filter(item => item.status.includes(filterType) && item.value.includes(searchField))
    }

    const mockOldItems: Items[] = [
        { value: 'item 1', status: 'done', itemIndex: 0 },
        { value: 'item 2', status: 'pending', itemIndex: 1 },
    ]
    let mockSearchField = ''
    let mockFilterType: FilterType = ''
    const mockSetSearchField = jest.fn();
    const mockSetFilterType = jest.fn();
    const mockSetItems = jest.fn();

    const addItemsWrapper = (filterType?: boolean) => shallow(<AddItems items={filterType ? [] : mockItems(mockOldItems, '', '')} oldItems={mockOldItems}
        filterType={filterType ? 'done' : mockFilterType} setFilterType={mockSetFilterType} searchField={mockSearchField} setSearchField={mockSetSearchField} setItems={mockSetItems} />)

    it('Should render AddItems component', () => {
        expect(addItemsWrapper()).toMatchSnapshot()
    })

    it('Button should be disabled false', () => {
        addItemsWrapper().find('[id="add-item-field"]').simulate("change", { target: { value: "Test" } })
        setTimeout(() => {
            expect(addItemsWrapper().find('[id="add-item-field"]').prop('value')).toStrictEqual("Test")
            expect(addItemsWrapper().find('[id="add-item-btn"]').props().disabled).toBe(false)
            addItemsWrapper().find('[id="add-item-btn"]').simulate('click')
            expect(mockSetItems).toBeCalledTimes(1)
        }, 500)
    })

    it('Filters should be cleaned', () => {
        const clearButton = addItemsWrapper(true).find('[id="clear-filter"]')
        clearButton.simulate('click')
        expect(mockSetFilterType).toBeCalledTimes(1)
    })
})