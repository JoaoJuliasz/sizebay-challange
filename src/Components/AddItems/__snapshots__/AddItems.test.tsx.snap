// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItems Should render AddItems component 1`] = `ShallowWrapper {}`;
